### PR TITLE
use blueprint-compiler snap

### PR DESCRIPTION
**Pull request by soumyaDghosh**

### CODE REVIEW

**Comment by cmkohnen** [comment]
This seams to break aarch64. Don't really know why though

**Comment by soumyaDghosh** [comment]
Exactly I am also not understanding the issue. Let's keep this open.

**Comment by cmkohnen** [comment]
Could this be an upstream issue? As for some reason the batch-compile command can not be found in CI

**Comment by soumyaDghosh** [comment]
The blueprint-compiler snap is maintained by me. So, as much I can check, no such issue. Still.. I will wait.

**Comment by cmkohnen** [comment]
Do you by chance have a link to that manifest on hand?

**Comment by soumyaDghosh** [comment]
Yup.
https://github.com/jameswestman/blueprint-compiler-snap

BTW, Tagger also uses batch-compile

**Comment by cmkohnen** [comment]
thanks, will try looking into that later today, maybe I can find the issue (it's probably our mess of a build script trying to get blueprint and vala play together)

**Comment by Sjoerd1993** [comment]
What I'm a bit confused about is that it fails only on arm64, while it works fine for x86, given it does the exact same thing on both architectures.

Given the error here, it seems to specifically go wrong with the `batch-compile` command indeed. Which is triggered in our meson file like this:
```
blueprints = custom_target('blueprints',
  input: blueprints_source,
  output: '.',
  command: [find_program('blueprint-compiler'), 'batch-compile', '@OUTPUT@', '@CURRENT_SOURCE_DIR@', '@INPUT@'],
)
```

From the lines in the log we get this:
```
:: [1/40] Generating po/tr/LC_MESSAGES/graphs-tr.mo with a custom command
:: [2/40] Generating po/sv/LC_MESSAGES/graphs-sv.mo with a custom command
:: [3/40] Generating po/nl/LC_MESSAGES/graphs-nl.mo with a custom command
:: [4/40] Generating data/blueprints with a custom command
:: FAILED: data
:: /snap/bin/blueprint-compiler batch-compile data/. ../src/data ../src/data/ui/add_equation_window.blp ../src/data/ui/add_style.blp ../src/data/ui/curve_fitting.blp ../src/data/ui/dialogs.blp ../src/data/ui/edit_item.blp ../src/data/ui/export_figure.blp ../src/data/ui/figure_settings.blp ../src/data/ui/fitting_parameters.blp ../src/data/ui/import.blp ../src/data/ui/help_overlay.blp ../src/data/ui/item_box.blp ../src/data/ui/smoothen_settings.blp ../src/data/ui/style_color_box.blp ../src/data/ui/style_editor.blp ../src/data/ui/style_preview.blp ../src/data/ui/transform_window.blp ../src/data/ui/window.blp
:: error: unknown command "batch-compile", see 'snap help'.
```

Compare this with the succesful run from the x86 snap:
```
[1/40] Generating po/nl/LC_MESSAGES/graphs-nl.mo with a custom command
[2/40] Generating po/sv/LC_MESSAGES/graphs-sv.mo with a custom command
[3/40] Generating po/tr/LC_MESSAGES/graphs-tr.mo with a custom command
[4/40] Generating data/se.sjoerd.Graphs.appdata.xml with a custom command
[5/40] Generating data/se.sjoerd.Graphs.desktop with a custom command
[6/40] Generating data/blueprints with a custom command
[7/40] Generating data/blueprints_hack with a custom command
```

It's very interesting that it seems to skip the appdata and desktop files. I wonder how that could happen, does it try to do this in a different order? The fact that it does find the `/snap/bin/blueprint-compiler` part, seems to tell me at least that it does find an installed program called `blueprint-compiler`. Then it should use `batch-compile` in this case as an argument for `blueprint-compiler`. This is all neatly defined in [Blueprints main.py:](https://gitlab.gnome.org/jwestman/blueprint-compiler/-/blob/main/blueprintcompiler/main.py) 
```
        batch_compile = self.add_subcommand(
            "batch-compile",
            "Compile many blueprint files at once",
            self.cmd_batch_compile,
        )
```

Yet, it clearly gives the above error. So it seems something went wrong with the `blueprint-compiler` install.

The thing that confuses me:
- Why does the log complain that it cannot find the command `batch-compile`? It would make sense to me that it would rather give a complaint that it's an invalid argument. This kinda reads like it tries to execute `batch-compile` specifically. Could it be that `blueprint-compiler` is thus not installed properly?
- What's with the `FAILED: data` line? Could it be that it cannot access this folder for some reason? Perhaps the way `blueprint-compiler` is installed makes it to not have access to the needed folder? Just a complete guess, it would not make that much sense though given that's no issue for the x86 build.

**Comment by soumyaDghosh** [comment]
@Sjoerd1993 @cmkohnen Don't waste your precious time on this. As I don't have an arm machine, I was not able to test locally. So, slowly I'll look into them.

**Comment by soumyaDghosh** [comment]
@Sjoerd1993 so, a small request. I think, we should replace the workflows now, because this is old/not supported, and probably that's reason behind arm failing. To ensure that arm is still building, try once in your machine by remote-build, run `snapcraft remote-build`. I tested and it worked. If it works for you too, I'll change the workflow

**Comment by Sjoerd1993** [comment]
> @Sjoerd1993 so, a small request. I think, we should replace the workflows now, because this is old/not supported, and probably that's reason behind arm failing. To ensure that arm is still building, try once in your machine by remote-build, run `snapcraft remote-build`. I tested and it worked. If it works for you too, I'll change the workflow

Yeah it definitely feels like we should replace the Snap workflow. I quickly looked at what [Tagger does](https://github.com/NickvisionApps/Tagger/blob/main/.github/workflows/snap.yml). That may work, but they skip amd64, not sure if there's a reason for that. 

In the worst case we can also check snap only for amd64 and let it fail for arm64 in the Snap store. Since we're not shipping ready made Snap files that would still work be okay, as the Snap build system ensures that those still work. Yet it would be preferred to do the amd64 check in the workflow.

**Comment by soumyaDghosh** [comment]
@Sjoerd1993 Kindly add this in the github env as `LP_BUILD_SECRET`. I have mailed the details to you. Kindly check.

**Comment by Sjoerd1993** [comment]
I tried to add the credentials you sent me in a new `LP_BUILD_SECRET` environment. I added them as Environment secret, which are described in the description as:

```
Environment secrets

Secrets are encrypted environment variables. They are accessible only by GitHub Actions in the context of this environment. 
```
Something is not working yet. Not sure if I should use `Environment Variable` instead, which seems to be the same thing but not encrypted. But won't try that unless you specifically say so, as I am not 100% sure if they'd be readable from the outside logs then.

**Comment by soumyaDghosh** [comment]
> I tried to add the credentials you sent me in a new `LP_BUILD_SECRET` environment. I added them as Environment secret, which are described in the description as:
> 
> ```
> Environment secrets
> 
> Secrets are encrypted environment variables. They are accessible only by GitHub Actions in the context of this environment. 
> ```
> 
> Something is not working yet. Not sure if I should use `Environment Variable` instead, which seems to be the same thing but not encrypted. But won't try that unless you specifically say so, as I am not 100% sure if they'd be readable from the outside logs then.

Check the mail again..

**Comment by soumyaDghosh** [comment]
Don't merge this PR just now. I will test some fails tomorrow, and add some more features in workflow. Let's merge all of them together.

**Comment by Sjoerd1993** [comment]
Sounds good to me. Also, maybe it makes sense to remove the snap patch in this PR instead of the other one.

**Comment by Sjoerd1993** [comment]
> Don't merge this PR just now. I will test some fails tomorrow, and add some more features in workflow. Let's merge all of them together.

Just let us know when this is ready for review/merging. Just so that we know when to look at it, there's no immediate rush to get it merged :)

And thanks again for the work on the snap package. Your efforts really make it worthwhile for us to also have it on Snap as well. Right now we're at 470 active installs. User numbers have been somewhat stable last month, but I think we can get a nice boost in numbers after next release which is going to be a very major one.

**Comment by soumyaDghosh** [comment]
Why is the workflow not working again?

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -79,6 +71,8 @@ parts:
     build-environment:
       - PYTHONPATH: $CRAFT_STAGE/usr/lib/python3/dist-packages:$PYTHONPATH
       - GI_TYPELIB_PATH: /snap/gnome-42-2204-sdk/current/usr/lib/$CRAFT_ARCH_TRIPLET/girepository-1.0:/snap/gnome-42-2204-sdk/current/usr/lib/girepository-1.0
+    build-snaps:
+      - blueprint-compiler/latest/edge
     override-pull: |
       craftctl default
       patch -p1 < $CRAFT_PROJECT_DIR/snap/graphs.patch
```
[the remark this answers is not in the document]
Comment (soumyaDghosh) [reply snap/snapcraft.yaml:78]: Yeah sure.. I am currently experimenting with the github ci. I guess I have fixed it and made it foolproof. The next commit will remove the patch

**Comment by soumyaDghosh** [comment]
@Sjoerd1993 with this artifact of snap, will it be okay?

**Comment by Sjoerd1993** [comment]
> @Sjoerd1993 with this artifact of snap, will it be okay?

From what I can tell, it seems to upload the correct Snap file. But as far as I can tell it is not currently being used after doing that. Unless we'll automatically push to the edge channel in the Snap store, I'm not sure if there's a point in uploading said artifact.

**Comment by soumyaDghosh** [comment]
If you want I can do that too. That's actually a great idea, but be aware with the release one, trigger one more build in the store, to be sure that it's the stable release. BTW, now I am going to push a commit that should fail the CI. You can decide which way to go,

- keep the current way. In the current way too, the latest build is in the edge channel
- Go the github ci method. During release, it may be confusing as the version number is the same.(we can tackle this with a temp release number 1.7.0-beta until real release comes out)

**Comment by soumyaDghosh** [comment]
Okay, so the CI is working properly. With throwing errors when it cannot found the snap file that's built. So, I am just removing the upload artifact step.

**Comment by Sjoerd1993** [comment]
My guess is that it may now needs the `upload-artifact` anyway step to publish to the edge channel.

**Comment by soumyaDghosh** [comment]
Nope, it needs another environment variable the store login key, add it as `STORE_LOGIN`.

https://github.com/snapcore/action-publish/#store-login

regarding what's wrong with the CI, probably the launchpad issue. Just restarting the CI.

**Review by Sjoerd1993 — APPROVED** [review]
I've added a key to the Github secrets according to [the instructions](https://github.com/snapcore/action-publish/#store-login). So we'll see if it triggers correctly when pushing to edge. But overall this seems to work well. 

Again, thanks a lot for all your work with the Snap packages. I see you've also been working on the blueprint snap, which also makes the entire process cleaner instead of having to compile that here.

**Comment by soumyaDghosh** [comment]
Ready to merge. The last commit was to use blueprint-compiler from latest/stable instead of edge.

**Comment by cmkohnen** [comment]
Just a quick question: Is it possible to specify a tag vor the blueprint-compiler? As future releases might still break backwards-compatibility.

**Pull request merged**